stl/inc/format

```diff
@@ -445,103 +449,239 @@ _NODISCARD constexpr bool _Is_execution_charset_utf8() {
-        return -1;
-    } else { // next code unit is '\0'
-        return 1;
-    }
-}
+        const auto _Len = static_cast<size_t>(_Last - _First);
 
-_NODISCARD inline int _Code_units_in_next_character(const char* _First, const char* _Last, const _Cvtvec& _Cvt) {
-    // Returns a count of the number of code units that compose the first encoded character in
-    // [_First, _Last), or -1 if [_First, _Last) doesn't contain an entire encoded character or
-    // *_First is not a valid lead byte.
-    _STL_INTERNAL_CHECK(_First < _Last);
+        if (_Ch < 0b1110'0000u) {
+            // check for non-lead byte or partial 2-byte encoded character
+            return (_Ch >= 0b1100'0000u && _Len >= 2) ? 2 : -1;
+        }
 
-    if constexpr (_Is_execution_charset_utf8_v) {
-        return _Utf8_code_units_in_next_character(_First, _Last);
-    } else {
-        switch (_Cvt._Mbcurmax) {
+        if (_Ch < 0b1111'0000u) {
+            // check for partial 3-byte encoded character
+            return (_Len >= 3) ? 3 : -1;
+        }
+
+        // check for partial 4-byte encoded character
+        return (_Len >= 4) ? 4 : -1;
+    }
+
+    _NODISCARD static constexpr int _Estimate_utf8_character_width(const char* const _Ptr, const int _Units) noexcept {
+        // Return an estimate for the width of the character composed of _Units code units,
+        // whose first code unit is denoted by _Ptr.
+        auto _Ch = static_cast<char32_t>(*_Ptr);
+        switch (_Units) {
         default:
-            _STL_INTERNAL_CHECK(!"Bad number of encoding units for this code page");
-            [[fallthrough]];
         case 1:
-            return 1; // all characters have only one code unit
-
         case 2:
-            return _Double_byte_encoding_code_units_in_next_character(_First, _Last, _Cvt);
+            return 1;
+        case 3:
+            _Ch &= 0b1111u;
+            break;
+        case 4:
+            _Ch &= 0b111u;
+            break;
+        }
+
+        for (int _Idx = 1; _Idx < _Units; ++_Idx) {
+            _Ch = _Ch << 6 | (_Ptr[_Idx] & 0b11'1111u);
+        }
 
-        case 4: // Assume UTF-8 (as does _Mbrtowc)
+        if (_Units == 3) {
+            return _Unicode_width_estimate<_Width_estimate_low_intervals>(_Ch);
+        }
+
+        return _Unicode_width_estimate<_Width_estimate_high_intervals>(_Ch);
+    }
+
+public:
+    _NODISCARD int _Units_in_next_character(const char* _First, const char* _Last) const noexcept {
+        // Returns a count of the number of code units that compose the first encoded character in
+        // [_First, _Last), or -1 if [_First, _Last) doesn't contain an entire encoded character or
+        // *_First is not a valid lead byte.
+        _STL_INTERNAL_CHECK(_First < _Last);
+
+        if constexpr (_Statically_Utf8) {
             return _Utf8_code_units_in_next_character(_First, _Last);
+        } else {
+            switch (this->_Cvt._Mbcurmax) {

```

Comment:
I'm not sure what you're referring to?

Reply:
I believe there was a question in discord about that https://discord.com/channels/737189251069771789/805902947682484244/831634999336763432

Reply:
Oh, yes! The "why is the compiler complaining about access to a non-member in a discarded statement" thing. Compilers are bad. [stmt.if]/2 says:

> ... During the instantiation of an enclosing templated entity ([temp.pre]), if the condition is not value-dependent after its instantiation, the discarded substatement (if any) is not instantiated.

but both Clang and MSVC behave as if it said "During the instantiation of an enclosing templated entity ([temp.pre]), if the condition is **value-dependent before its instantation, but** not value-dependent after its instantiation, the discarded substatement (if any) is not instantiated." Fix was to template this class on a `bool` that defaults to `_Is_execution_charset_utf8` and using that argument for `if constexpr` instead of making it non-template and `if constexpr`ing directly on `_Is_execution_charset_utf8_v` (which variable is no longer necessary).

Reply:
FWIW, I haven't reported compiler bugs for this behavior since I'm 99% certain that a Clang developer who shall remain nameless (who was once C++ Project Editor and has a GitHub ID that rhymes with mygoloid) will tell me that is the intent of the Standard and file an editorial issue to clarify the wording. Indeed, the wording generally tries to allow compilers to substitute template arguments for parameters throughout an entire template body at instantiation time to enable early diagnostics. We library folk just barely won the battle to delay substitution into associated constraints until overload resolution time, I don't want to push my luck ;)